oidc/clientassertion/client_assertion.go

```diff
@@ -0,0 +1,206 @@
+package clientassertion
+
+import (
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-jose/go-jose/v3"
+	"github.com/go-jose/go-jose/v3/jwt"
+	"github.com/hashicorp/go-uuid"
+)
+
+const (
+	// ClientAssertionJWTType is the proper value for client_assertion_type.
+	// https://www.rfc-editor.org/rfc/rfc7523.html#section-2.2
+	ClientAssertionJWTType = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"
+)
+
+var (
+	// these may happen due to user error
+	ErrMissingClientID    = errors.New("missing client ID")
+	ErrMissingAudience    = errors.New("missing audience")
+	ErrMissingAlgorithm   = errors.New("missing signing algorithm")
+	ErrMissingKeyOrSecret = errors.New("missing private key or client secret")
+	ErrBothKeyAndSecret   = errors.New("both private key and client secret provided")
+	// if these happen, either the user directly instantiated &ClientAssertion{}
+	// or there's a bug somewhere.
+	ErrMissingFuncIDGenerator = errors.New("missing IDgen func; please use New()")
+	ErrMissingFuncNow         = errors.New("missing now func; please use New()")
+)
+
+// New sets up a new ClientAssertion to sign private key JWTs
+func New(clientID string, audience []string, opts ...Option) (*ClientAssertion, error) {
+	a := &ClientAssertion{
+		clientID: clientID,
+		audience: audience,
+		headers:  make(map[string]string),
+		genID:    uuid.GenerateUUID,
+		now:      time.Now,
+	}
+	for _, opt := range opts {
+		opt(a)
+	}
+	if err := a.Validate(); err != nil {
+		return nil, fmt.Errorf("new client assertion validation error: %w", err)
+	}
+	return a, nil
+}
+
+// ClientAssertion signs a JWT with either a private key or a secret
+type ClientAssertion struct {
```

Comment:
This name stutters a bit (`clientassertion.ClientAssertion`). What do you think about calling this `JWT` instead? Since the only exposed functions are `Validate` and `SignedToken`, we could just rename `SignedToken` to `Serialize` or something.